### PR TITLE
Add Bool type, Assert and If expressions

### PR DESCRIPTION
**Pull request by fnoble**

### CODE REVIEW

File: src/Plover/Reduce.hs [hunk 1]
```diff
@@ -290,10 +298,22 @@ typeCheck' (Vec var bound body) = do
...
+  btype <- typeCheck' c
+  case btype of
+    BoolType -> return BoolType
+    t -> error $ "Assert condition must be BoolType, got: " ++ sep t btype
+typeCheck' (Equal a b) = do
+  atype <- typeCheck' a
+  btype <- typeCheck' b
+  if atype == btype
+    then return BoolType
+    else error $ "Equality test between values with different types: " ++ sep atype btype
```
Comment (kovach) [inline src/Plover/Reduce.hs]: these 'error' calls should be 'left'

File: src/Plover/Reduce.hs [hunk 2]
```diff
@@ -212,6 +213,13 @@ typeCheck' e@(a :<= b) = do
   tr <- typeCheck' b
   unifyGen tl tr
   return Void
+typeCheck' e@(If cond t f) = do
+  local $ typeCheck' t
+  local $ typeCheck' f
+  ctype <- typeCheck' cond
+  case ctype of
+    BoolType -> return Void
```
Comment (kovach) [inline src/Plover/Reduce.hs]: I think this could be an issue if we 'Return' from inside a branch

File: src/Plover/Reduce.hs [hunk 3]
```diff
@@ -720,6 +744,29 @@ compileStep' ctxt (a :> b) = do
...
+  return (If c' t' f')
+
+-- Vector comparison
+compileStep' _ e@(x :=: y) = do
+  tx <- local $ typeCheck x
+  ty <- local $ typeCheck y
+  case (tx, ty) of
+    -- pointwise comparison
+    (VecType (len1 : _) t1, VecType (len2 : _) t2) -> do
+      error "Vector comparison not implemented"
```
Comment (kovach) [inline src/Plover/Reduce.hs]: -> left

File: src/Plover/Reduce.hs (hunk 1 above)
Comment (fnoble) [inline src/Plover/Reduce.hs]: `typeCheck` currently uses `error` everywhere, should I change them all to `left`?

File: src/Plover/Reduce.hs (hunk 3 above)
Comment (fnoble) [inline src/Plover/Reduce.hs]: `compileStep'` currently uses `error` everywhere, should I change them all to `left`?

File: src/Plover/Reduce.hs (hunk 2 above)
Comment (fnoble) [inline src/Plover/Reduce.hs]: Whats the solution? Make sure both branches have the same type and then make the if statement also inherit that type?

File: src/Plover/Reduce.hs (hunk 1 above)
Comment (kovach) [inline src/Plover/Reduce.hs]: it really only uses error in one place (in a ! b that call should be removed), and left everywhere else for type checking errors

File: src/Plover/Reduce.hs (hunk 3 above)
Comment (kovach) [inline src/Plover/Reduce.hs]: yeah, error in compileStep is fine

File: src/Plover/Reduce.hs [hunk 4]
```diff
@@ -347,13 +367,19 @@ typeCheck' (s :. f) = do
...
+    PtrType t -> do
+      t' <- normalizeType t
+      case t' of
+        StructType _ (ST _ fieldTypes) -> do
+          case lookup f fieldTypes of
+            Just t'' -> return t''
+            Nothing -> left $ "typeCheck'. field is not a member of struct:\n"
+              ++ sep f fieldTypes
+        _ -> error $ "typeCheck'. not a pointer to a struct: " ++ sep s t'
+    _ -> error $ "typeCheck'. not a pointer to a struct: " ++ sep s t
```
Comment (kovach) [inline src/Plover/Reduce.hs]: left

**Comment by fnoble** [comment]
@kovach : ready for review again

**Pull request merged**